`make O=build` out-of-tree

vtjnash opened the pull request:
this implements out-of-tree builds (without requiring cmake :smile:)

inspired by the linux kernel kbuild system, this uses `make O=` to setup a target directory. it then creates dummy makefiles in the out-of-tree build directories so that the user can more easily do incremental build work. the user can place independent `Make.user` files in each build tree to override the global Make.user file in the toplevel source directory.

additionally, it moves all source-file related content into deps/srccache for easier tracking (and build sharing).

vtjnash committed:
fix makefile rule for copying examples to dest

ref #8428 (comments). fix #12539

vtjnash committed:
try to avoid a possible bug in gnumake

some AppVeyor builds (like https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.7584/job/2yoa82naaj9xfhyl) seem to be getting a corrupt expansion.
this is more accurate at being a recursive wildcard anyways

vtjnash committed:
support out-of-tree builds in each subdirectories

inspired by KBuild, supporting specification of both (or either)
the input (SRCDIR) and intermediary (BUILDDIR) directories:
mkdir -p build/src && cd build/src && \
    make -f ../src/Makefile all && \
    make -C ../src/Makefile BUILDDIR=`pwd` all

vtjnash committed:
support O= output directory on toplevel make

respects both $(JULIAHOME)/Make.user and $(O)/Make.user

tkelman commented:
-1, would rather get further on cmake which just does this for you. If this can be made to work without breaking the CI, then maybe, but right now it's adding even more complication to the makefiles and seems like a really strange thing to be working on right this second.

yuyichao commented on base/fs.jl:
```diff
@@ -47,7 +47,7 @@ export File,
             _sizeof_uv_fs, uv_error
 
-include("file_constants.jl")
+include(string(length(Core.ARGS)>=2?Core.ARGS[2]:"","file_constants.jl"))  # include($BUILDROOT/base/file_constants.jl)
```
Well, this is a bit that is necessary in the cmake port for proper out-of-tree build as well. I'm currently adding this capability using a fallback path (which can be removed with stringreplace on the binary) in `jl_load` (the main advantage being not having to change anything in the julia code).

I'm at least interested in which way people prefer.

tkelman replied:
Why not think about populating these lists of constants in some other way? There's some really godawfully messy perl going on in the makefiles right now to generate these, no one would miss those lines if they went away.

vtjnash replied:
i would rather stick with the minimal change (this) in this PR, but can certainly entertain other ideas for implementation in a separate PR. the most logical change would probably to change this to a ccall to a method that has a big case statement to pick the right answer for the input string.

tkelman replied:
I still don't get what is going on here, or why some of these need `vcat` and `.data`. Let's not rush this with ugly confusing code.

vtjnash replied:
some of these are early enough that strings are defined yet, so they need `vcat` and `.data`. it should be possible to make that more opaque in a future PR, but that would take a bit more restructuring and is unrelated to this PR. (this change, or something equivalent also needs to go in to make the cmake build work, so it isn't wasted work)

tkelman commented on base/Makefile:
```diff
@@ -19,22 +21,22 @@ else
-fenv_constants.jl: ../src/fenv_constants.h
-	@$(PRINT_PERL) $(CPP_STDOUT) -DJULIA -I../deps/openlibm/include ../src/fenv_constants.h | tail -n 8 | perl -ple 's/\sFE_UN\w+/ 0x10/g; s/\sFE_O\w+/ 0x08/g; s/\sFE_DI\w+/ 0x04/g; s/\sFE_INV\w+/ 0x01/g; s/\sFE_TON\w+/ 0x00/g; s/\sFE_UP\w+/ 0x800/g; s/\sFE_DO\w+/ 0x400/g; s/\sFE_TOW\w+/ 0xc00/g' > $@
+$(BUILDDIR)/fenv_constants.jl: $(SRCDIR)/../src/fenv_constants.h
+	@$(PRINT_PERL) $(CPP_STDOUT) -DJULIA -I$(SRCDIR)/../deps/openlibm/include $< | tail -n 8 | perl -ple 's/\sFE_UN\w+/ 0x10/g; s/\sFE_O\w+/ 0x08/g; s/\sFE_DI\w+/ 0x04/g; s/\sFE_INV\w+/ 0x01/g; s/\sFE_TON\w+/ 0x00/g; s/\sFE_UP\w+/ 0x800/g; s/\sFE_DO\w+/ 0x400/g; s/\sFE_TOW\w+/ 0xc00/g' > $@
```
the openlibm path strikes me as though it would be broken in a NO_GIT build - do we need it?

vtjnash replied:
only on windows, IIRC

tkelman replied:
Oh, whoops, I added that in #6230. So only with MSVC I think, which I believe is broken in a NO_GIT build. Possibly for reasons other than this too.

vtjnash committed:
cleanup old, unused cross-compile code that depended on gnu extensions to xargs

vtjnash commented:
> If this can be made to work without breaking the CI

Of course (also done now)

> would rather get further on cmake which just does this for you

sure. the benefit in my mind here is that we can merge this now (since it doesn't change the build system much to add), and keep working on the cmake rewrite

tkelman commented:
Would rather wait until after branching release-0.4 to merge this. It should theoretically be non-disruptive, but some of these build configurations don't get tested very often.

yuyichao commented:
FWIW, I do prefer merging the necessary bits for out-of-source build of sysimg first. It has been the part that introduce the most conflict...

vtjnash commented:
> Would rather wait until after branching release-0.4 to merge this. It should theoretically be non-disruptive, but some of these build configurations don't get tested very often.

I'm actually thinking that if we can get this in before v0.4, it'll make it easier to keep a build tree around for v0.4, so that developers can test against many branches more easily.

aside from moving a lot of source files out of deps into deps/srccache, the way i implemented these build options, they are usually set to the empty string, so that this has very minimal impact on the in-tree build. 

the src directories have supported this option for quite some time, this just finishes that work by extending support to deps, ui, base, etc.

tkelman commented:
The `ARGS` thing feels really hacky, and this breaks Cygwin cross-compile. This is supposed to be minimal impact, but it isn't as minimal as you think it is. It needs to be reviewed and tested, and we're trying to get master into an RC state. I don't think this is high enough priority to worry about right now.

vtjnash commented:
> The ARGS thing feels really hacky, and this breaks Cygwin cross-compile

thanks for point that out. i think i've fixed it (still waiting for my cross-compile build of llvm to finish): https://github.com/JuliaLang/julia/commit/4bba411bcf3558e55c8cfb31c3fd8ab56aba5aa7#diff-b67911656ef5d18c4ae36cb6741b7965R193

tkelman commented:
Are you cross compiling from cygwin or linux? The line you pointed to isn't the issue, it's the failure to find the stdlibs. You're undoing part of #5965, the difference now is breaking the Cygwin cross-compile means no Windows binaries.

vtjnash commented:
> Are you cross compiling from cygwin or linux? The line you pointed to isn't the issue, it's the failure to find the stdlibs. You're undoing part of #5965, the difference now is breaking the Cygwin cross-compile means no Windows binaries.

my cygwin build is further behind. i didn't realize the mingw64-cygwin compiler was lying about it's paths, i've put that line back.

vtjnash committed:
prefer shelling out to the llvm-config-host binary, rather than trying to indirectly execute the target platform llvm-config

tkelman commented:
LLVM patches are still broken on cygwin here
```
/bin/sh: -c: line 0: syntax error near unexpected token `newline'
/bin/sh: -c: line 0: `cd /home/Tony/julia/deps/srccache/llvm-3.3 && patch -p1 < '
Makefile:669: recipe for target '/home/Tony/julia/deps/srccache/llvm-3.3/llvm-3.3.patch-applied' failed
make[1]: *** [/home/Tony/julia/deps/srccache/llvm-3.3/llvm-3.3.patch-applied] Error 1
Makefile:60: recipe for target 'julia-deps' failed
```

vtjnash committed:
fix application of llvm patches

vtjnash commented:
ah, that didn't actually work anywhere.

tkelman commented:
still doesn't?
```
/bin/sh: line 0: cd: /home/Tony/julia/deps/srccache/llvm-3.3: No such file or directory
Makefile:669: recipe for target '/home/Tony/julia/deps/srccache/llvm-3.3/llvm-3.3.patch-applied' failed
make[1]: *** [/home/Tony/julia/deps/srccache/llvm-3.3/llvm-3.3.patch-applied] Error 1
Makefile:60: recipe for target 'julia-deps' failed
make: *** [julia-deps] Error 2
```

vtjnash commented:
oh, right. i guess it would also help if the patch files depended on the source actually existing too.
thanks for checking all of these.

tkelman commented on deps/Makefile:
```diff
@@ -1957,26 +2002,28 @@ LIBGIT2_OPTS += -DCMAKE_FIND_ROOT_PATH=/usr/$(XC_HOST) -DCMAKE_FIND_ROOT_PATH_MO
-$(LIBGIT2_OBJ_SOURCE): $(LIBGIT2_SRC_DIR)/build/Makefile
-	$(MAKE) -C $(LIBGIT2_SRC_DIR)/build
+$(LIBGIT2_OBJ_SOURCE): $(BUILDDIR)/$(LIBGIT2_SRC_DIR)/Makefile
+	$(CMAKE) --build $(dir $<)
```
last I checked this has issues with parallel make

vtjnash replied:
we may need to look into that further than. this change was already done for llvm and was being done here to be consistent across all cmake usage in the Makefile. it is in support of the following comment: https://github.com/JuliaLang/julia/pull/12463/files#diff-3ba529ae517f6b57803af0502f52a40bR57

tkelman replied:
That link doesn't actually point anywhere for me. Using this form can help a bit if you're using different generators like ninja, but you know what would help with that even more? Using cmake for as much of our build as possible. gmake doesn't know that `$(CMAKE) --build` will usually be a recursive make call.

vtjnash replied:
that's a somewhat faulty argument. I could just as well say that changing the build rules for libgit2 to make (which would be much easier than the reverse) would fix the issue with recursion here.

tkelman replied:
Sure, but there's a PR open that gets much of the way towards building Julia with cmake, I know of no such work to make generic cross-platform gnu makefiles for libgit2. We don't need to use the same generator for all cmake dependencies, at the moment it makes sense to me to always use the makefile generator for libgit2 since we're very rarely building it incrementally. Keno builds LLVM incrementally fairly often I imagine.

tkelman commented:
I'm still *really* iffy on merging this before we branch. I don't think it's worth the risk. There are exotic configurations or rarely used dependencies that this could easily be breaking in subtle ways that wouldn't get reported for some time. And merging this would force everyone to re-download and re-configure every dependency (and re-build in several cases, fftw is re-running its entire make here), which won't go over well.

What's going on with `ARGS`? And `.data`? That looks really bizarre and confusing. Not a confidence-inspiring PR for RC time.

tkelman commented:
after 50 minutes of rebuilding all the autotools deps,

```
    CC src/support/_setjmp.win64.o
    CC src/support/_longjmp.win64.o
    LINK src/support/libsupport.a
make[1]: *** No rule to make target 'disasm.o', needed by '/home/Tony/julia/usr/bin/libjulia.dll'. Stop.
make[1]: *** Waiting for unfinished jobs....
    CC src/module.o
Makefile:69: recipe for target 'julia-src-release' failed
make: *** [julia-src-release] Error 2
```

StefanKarpinski commented:
Agree: let's not merge this now.

vtjnash commented:
> There are exotic configurations or rarely used dependencies that this could easily be breaking in subtle ways that wouldn't get reported for some time

yes, but on the plus side, i would actually be able to test a large build matrix before pushing a PR (win32/64, native32/64, llvm3.3/3.6/svn). it may even be possible to add ARM to that list. that's why i want this so much.

> No rule to make target 'disasm.o',

this looks like a race condition somewhere. it means that it could not satisfy something in the dependency list: `disasm.cpp $(HEADERS) $(shell which $(LLVM_CONFIG))`. did it complete if you tried `make` again? if so, i think that may have been preexisting, but i'll try to push a fix for it anyways (the fix should also help reduce compile times on cygwin very slightly).

vtjnash committed:
switch to := in Makefiles almost exclusively

as per recommendation of Chapter 9 "Makefile language" of the
developer information for linux kernel makefiles:
https://www.kernel.org/doc/Documentation/kbuild/makefiles.txt

tkelman commented:
You don't have to convince me of why you want this. I just don't want it right now. You're not testing on freebsd, and you're only testing cross-builds for Windows from a single tree. Use Vagrant and multiple checkouts for 0.4, it's been working. We can make this work during 0.5-dev, but preferably with a cleaner way of dealing with the generated Julia sources. This is not RC ready, and there are better things to work on right now - would be great to have more people trying to figure out how to get the docs in shape, first and foremost.

Trying `make -j2` multiple times repeatedly executes mpfr's make (so that's getting unfixed here), gives the "No rule to make target disasm.o" error repeatedly, and after a few more times also gives a "io-error: file: could not open /home/Tony/julia/src/mk_julia_flisp_boot.scm" - you're missing a cygpath_w there.

vtjnash committed:
add missing touch rules to prevent gmp, mpfr, and libuv from attempting to rebuild repeatedly

vtjnash committed:
add missing cygpath_w conversion for mk_julia_flisp_boot

tkelman commented:
Something seems to have broken the checksumming. Possibly the massive change to `:=` which could easily be breaking the way we set make variables from the environment, command line, and/or `Make.user` on CI (not to mention local machines).

We're literally documentation-epsilon away from an RC. Not the time for large build system changes.

vtjnash commented:
> Trying make -j2 multiple times repeatedly executes mpfr's gmp's make (so that's getting unfixed here), 

those two bugs are actually pre-existing, sorry. i'll add the fix for gmp here too though.

vtjnash commented:
> You don't have to convince me of why you want this. I just don't want it right now

i want this because i break the build rather often for all of the platforms except the one i'm sitting in front of.

i want this now because i have often wished during backporting to have the ability to simultaneously verify that the patch works on all configurations. and while RC-1 might be tagged soon, there isn't exactly a dearth of bugs that are going to need to be addressed before getting to the actual release https://github.com/JuliaLang/julia/issues?q=is%3Aopen+is%3Aissue+milestone%3A0.4.x

tkelman commented:
The `disasm.o` problem is still present here, and more `cygpath_w` invocations are necessary for `bin2hex.scm` and probably every other file that flisp calls. None of this will be visible on appveyor since appveyor is building from MSYS1, and with a rather old version of make. It's a bit too time-consuming and large to install and cache cygwin on the appveyor workers, but if you have a login to the buildbots you can try forcing builds there of sha's from this branch to verify whether it'll work.

My point is you're never going to be able to check every single build from a single git clone. It's a pipe dream, give it up, you need to actually test the different build configurations. Out-of-tree builds will help a bit for those cases that can be done simultaneously from the same machine, but that's only ever going to be a minor subset of supported build configurations. And you still haven't gotten them all working here. Your time is better spent elsewhere right now.

vtjnash commented:
> My point is you're never going to be able to check every single build from a single git clone

that's not my goal at all. my goal is to be able to check that an order of magnitude more build configurations don't have syntax errors and be able to jump around between various branches more easily. the point of this change is not to be be able to better test the build infrastructure, the point is to be better able to test the code.

> The disasm.o problem is still present here

Are you using an installed system llvm binary? I might need to add a special case to consider it OK to run the native platform llvm-config.exe, instead of trying to run llvm-config-host. Do you have anything else in your Make.user file? Does `make debuginfo.o` work?

vtjnash commented:
> Are you using an installed system llvm binary? I might need to add a special case to consider it OK to run the native platform llvm-config.exe, instead of trying to run llvm-config-host. Do you have anything else in your Make.user file? Does make debuginfo.o work?

nvm. this is just good ol' cygwin / windows lying about file names / extensions. the fix should be simple enough

tkelman commented:
Again, I like and appreciate and agree with the goal, but the implementation is highly disruptive as evidenced by the half-dozen rounds of things this has broken. This change is not ready for prime-time.

Yes you do break the build fairly often, and we make a point of fixing it. I'd rather have 0.4 be in a known working state, and be careful about doing backports to it through PR's that can go through CI rather than hastily pushing straight to the branch, which is where most of the uncaught breakage on master comes from. This PR is bad for the stability of 0.4. We've already given up on easy backporting in the case of the codegen rewrite. This seems like the same argument again, though I'm not on your side of it this time since we're closer to release now.

You're probably right about needing an `$(EXE)` somewhere to fix the `disasm.o` issue. Still several more `cygpath_w`'s needed for flisp though.

vtjnash committed:
fix cygwin build of cpp files that depend on llvm-config

tkelman commented on Make.inc:
```diff
@@ -562,34 +584,38 @@ endif
+LLVM_CONFIG := $(build_bindir)/llvm-config$(HOST_EXE)
 else
-LLVM_CONFIG=$(build_bindir)/llvm-config$(EXE)
+LLVM_CONFIG := $(build_bindir)/llvm-config-host$(HOST_EXE)
```
In the spirit of "this does too many things at once," how about taking out the llvm-config-host change since it's not at all necessary or interdependent on out-of-tree builds. I get that you're trying to work towards arm cross, but one thing at a time.

tkelman commented:
Actually we can't use `llvm-config-host` with Cygwin-cross on 3.3 anyway. It's buggy and thinks we have `-lz`, but we don't.

vtjnash committed:
add make configure target to setup out-of-tree build directory, and add mention in README

vtjnash committed:
add checksums for arpack-ng testA.mtx file

vtjnash committed:
use an emulated llvm-config in 3.3 and before

llvm-config-host describes the build target, not the host target on llvm <= 3.3, so we must use the (emulated) llvm-config binary.
on newer versions of llvm, however, we can avoid the need to call target code and use llvm-config-host directly

vtjnash commented:
alright, i think that should cover everything:
1) uses llvm-config-host only if LLVM_VER != 3.3
2) adds README comments
3) adds `make configure O=build`
4) may reduce libuv's desire to reinstall
5) use `BUILD_EXE` name for consistency
6) fix warning on testA.mtx file checksum

vtjnash committed:
fix error in expansion order of $(HOSTCC) = $(CC)

tkelman commented:
This is still huge and makes way too many changes at once (when github refuses to show the diff on `deps/Makefile`, you know something's wrong). Bound to break buildbots and surprise distro packagers in a bad way. I'm very much against merging, even if you do address all of the issues below.

- I'd prefer @yuyichao's version of adding the build dir to the load path to handle the generated jl files.

- Don't change the build invocation on libgit2, we're not going to use a non-make generator there and it's a pain to have that build serially.

- The big change to `:=` feels gratuitous and unnecessary right now. That's bound to rearrange the override precedence of `Make.user` or environment or command-line variable settings in a way that will surprise and break something.

- `# make check doesn't actually compile (missing prototype / implementation for getline)` needs a separate issue for utf8proc.

- You had a to-do about `test/ccalltest`.

And likely unrelated but not helping matters,

- The cygwin-cross build is broken with a new segfault during bootstrap in combinatorics.

vtjnash commented:
> I'd prefer @yuyichao's version of adding the build dir to the load path to handle the generated jl files.

`Core.include` should not be handling search paths.

> Don't change the build invocation on libgit2, we're not going to use a non-make generator there and it's a pain to have that build serially.

On linux, this still builds in parallel. I don't know whether cygwin is broken in this respect, but I've been burned frequently enough by `make -jN` on windows that I stay away from it now.

> The big change to := feels gratuitous and unnecessary right now. That's bound to rearrange the override precedence of Make.user or environment or command-line variable settings in a way that will surprise and break something.

Let's please cut down on the FUD. If you don't understand the difference between `:=` and `=`, then you should typically be using `:=`. There is no difference in precedence between these two assignment operators, it simply defines the order in which the arguments to the assignments are expanded.

> make check doesn't actually compile (missing prototype / implementation for getline) needs a separate issue for utf8proc.

yes, but this has nothing to do with this PR

> You had a to-do about test/ccalltest.

good memory, but it wasn't really essential to include right now, so i can continue that in a future PR

> This is going to be a pain to bisect across, it messes with pcre's build in strange ways after switching do a different branch.

i think i should give the moved build folders a new name to cause less conflict. the change is good to finally happen. it has been held back waiting for a good reason to force it upon people.

> FFTW is failing to install.

did you delete the deps/fftw folders? if not, this probably the same issue as the pcre build and i can take the same resolution action to resolve both.

> And likely unrelated but not helping matters,
The cygwin-cross build is broken with a new segfault during bootstrap in combinatorics.

if unrelated, then it doesn't have anything to do with this PR. in a separate issue, someone can investigate.

vtjnash committed:
always segregate deps build files from the source files and the legacy builds

tkelman commented:
> `Core.include` should not be handling search paths.

Sure, but the inclusion of the generated jl files would be considerably less ugly if the `ARGS` during sysimg build could be handled in one place, with a comment describing what's going on so people other than you understand how the code works, rather than modifying paths half a dozen different confusing times.

> On linux, this still builds in parallel.

Absolutely sure? You can do `make -C deps distclean-libgit2 && make -j4` and you get libgit2 building in parallel? I don't, on either Linux or Cygwin-cross. I don't see how that could happen since `cmake --build` has no way to know that it's being called from within an outer parallel make. Driving cmake from make is pretty backwards and unconventional, but it's what we're doing for now. There might be a cmake flag to get `cmake --build` to run in parallel but my searches so far are coming up empty.

> I've been burned frequently enough by `make -jN` on windows that I stay away from it now.

The bug where parallel make would freeze was only ever present in MSYS1.

> Let's please cut down on the FUD.

The reason for the FUD is that I'm afraid, uncertain, and doubtful that this is actually a 100% harmless change that won't break anything. If this had been super important for 0.4 it would have been on the roadmap, and should have been done a while ago so it could be on master and we'd have enough time to verify whether or not it causes any problems. There's a lot more risk of breakage here than advantage of making 0.4 and 0.5 easier to switch between. Separate checkouts is mildly annoying but it works, there are more important non-build-related bugs to fix right now than a risky attempt to make switching branches easier. This is a feature that you, me, and only a handful of other people are going to use. It can wait.

> did you delete the deps/fftw folders? if not, this probably the same issue as the pcre build and i can take the same resolution action to resolve both.

Yes, even from a pristine build of this branch. The buildbots showed the same when I tried to make them build this branch from scratch. http://buildbot.e.ip.saba.us:8010/builders/package_osx10.9-x64/builds/349/steps/make/logs/stdio (long log, see the very end) Have you been doing a completely clean build here?

> if unrelated, then it doesn't have anything to do with this PR.

Actually it appears to have been related to the pcre branch switching issues. Something was building incorrectly, or stale, or in a corrupted state. Went away after completely rebuilding master from scratch.

tkelman commented:
This also interacts a bit strangely with `make -C deps O=build01 install-fftw`, ideally that would just run the same thing as `make -C deps install-fftw` but installing into toplevel `build01` but it ends up going into `deps/build01`. Maybe the individual dependency manipulations would be needed less often if they're building out-of-tree, but openblas and a few others are not capable of out-of-tree builds.

vtjnash commented:
> This also interacts a bit strangely with make -C deps O=build01 install-fftw, ideally that would just run the same thing as make -C deps install-fftw but installing into toplevel build01 but it ends up going into deps/build01. Maybe the individual dependency manipulations would be needed less often if they're building out-of-tree, but openblas and a few others are not capable of out-of-tree builds.

this is simply an (understandable) misunderstanding of make. `make -C deps` immediately turns into `cd deps && make` (it is documented to be equivalent). OTOH, once you run `make O=build01 configure`, you can run `make -C build01/deps install-fftw` to do incremental build manipulation

vtjnash commented:
> Absolutely sure? You can do make -C deps distclean-libgit2 && make -j4 and you get libgit2 building in parallel? I don't, on either Linux or Cygwin-cross. I don't see how that could happen since cmake --build has no way to know that it's being called from within an outer parallel make. Driving cmake from make is pretty backwards and unconventional, but it's what we're doing for now. There might be a cmake flag to get cmake --build to run in parallel but my searches so far are coming up empty.

the command line pass-through occurs via environment variables (source: https://www.gnu.org/software/make/manual/html_node/Options_002fRecursion.html#Options_002fRecursion), so it doesn't matter how it gets called: it stills gets all the options it is supposed to get. however, since I'm not using the special syntax `$(MAKE)`, I need to prefix this line with `+` to make everything work just right (for unknown reasons, Mac didn't complain about this, while cygwin does).

vtjnash committed:
revert libgit2 build to using $(MAKE)

$(CMAKE) with recursive makefiles would require a + prefix to pass the right environment, but that's too cmake-target specific to be useful for generalization of the cmake usage

vtjnash committed:
also delete build files as part of the deps distclean targets

tkelman commented on Make.inc:
```diff
@@ -207,75 +224,80 @@ ifneq (,$(findstring MSYS,$(BUILD_OS)))
-fPIC =
 ifeq ($(BUILD_OS), WINNT)
-PATH := $(PATH):$(build_libdir):$(build_private_libdir):/c/Program Files/7-zip
+PATH := $(PATH):$(build_libdir):$(build_private_libdir):/c/Program Files/7-zip:$(JULIAHOME)/dist-extras
```
https://github.com/JuliaLang/julia/commit/a6bb699c3d017b7a25826d6ed07f508c5076d99c#commitcomment-12605919

vtjnash replied:
i saw. they are appended to the end of PATH, so there's really no reason not to try to use them (if for some reason the others don't exist)

tkelman replied:
if we don't absolutely need to append to path it's better to leave it alone

vtjnash replied:
we may be able to just delete this line entirely. i don't think it is getting used anywhere anymore.

tkelman replied:
No, considering the only dll in either `build_libdir` or `build_private_libdir` is sys.dll I highly doubt these are doing anything. The addition of 7zip here is also quite old, and unpleasantly hardcoded - https://github.com/JuliaLang/julia/commit/c2323fbd25718707a4b9e2318d83ac16689da04b. Cygwin/msys2 git can act up if your path gets too long, so would be good to drop this.

tkelman replied:
Bump. Let's delete this line.

vtjnash committed:
perform basic validation of the user paths to ensure they don't contain shell metacharacters

vtjnash committed:
add a makefile trick for printing any makefile variable via `make print-VAR`

tkelman commented on Make.inc:
```diff
@@ -55,42 +56,74 @@ USEIFC  ?= 0
+# Make sure the user didn't try to specify a path that will confuse the shell / make
+METACHARACTERS := ][?*{}() $$%:;&|!\#,\\`\":
+ifneq (,$(findstring ',$(value BUILDROOT)))
+$(error cowardly refusing to build into directory with a shell-metacharacter in the path)
```
lets try to avoid unfamiliar jargon in error messages, shell metacharacter is not terminology that most people are going to recognize

vtjnash replied:
fortunately, there's google, which is quite happy to give you a list of these characters (this list is still missing a few like `<` `>` and `\n`, but i think this'll suffice since it doesn't actually seem to have bit anyone yet)

tkelman replied:
sure but if you can explain what the problem is in simple words without forcing someone to go to google, then just do the obvious version

vtjnash committed:
fix repackage_system_suitesparse4.make file to accommodate verification
that no shell metacharacters are present in the definition of JULIAHOME

tkelman commented:
Apparently the arpack tests don't link correctly with #8734, not too surprising since they've been commented out for so long.

```
dnsimp.o: In function `av_':
dnsimp.f:(.text+0x4f): undefined reference to `dgemv_'
collect2: error: ld returned 1 exit status
make[4]: *** [dnsimp] Error 1
make[4]: Leaving directory `/home/tkelman/Julia/julia-oot/bld01_linux64/deps/arpack-ng-3.2.0/TESTS'
make[3]: *** [check-am] Error 2
make[3]: Leaving directory `/home/tkelman/Julia/julia-oot/bld01_linux64/deps/arpack-ng-3.2.0/TESTS'
make[2]: *** [check-recursive] Error 1
make[2]: Leaving directory `/home/tkelman/Julia/julia-oot/bld01_linux64/deps/arpack-ng-3.2.0'
make[1]: *** [arpack-ng-3.2.0/checked] Error 2
```

vtjnash committed:
fixup arpack tests to support 64-bit blas

vtjnash commented:
> add GEMV to this list

thanks for point that out. that helped me find the rest of the issues more quickly.

unfortunately, one of the arpack tests is computing the wrong answer. i've pushed a new commit to turn off the tests when using the 64-bit interface and will open a new issue once this is merged so that someone can investigate that.

StefanKarpinski commented:
I agree with @tkelman that this seems like way too large and risky a change to attempt right before a release. Once again, the timing is a bit unfortunate. Maybe as soon as we release 0.4 something like this can be done and then once the kinks are worked out, we can backport it to the 0.4 branch.

vtjnash commented:
> this seems like way too large and risky a change to attempt right before a release. 

the code to support this change was merged much earlier in v0.4. it already works, it's just harder to use than necessary. this just finishes the (largely mechanical) translation of a few more Makefiles, and cleans up a few more makefile inconsistencies along the way to make it more useable.

unlike, say, switching to cmake, this has a really minor impact on the behavior of the build system.

StefanKarpinski commented:
It's 21 commits and 1700 lines of changed code. Calling this "minor" is crazy talk.

vtjnash commented:
it's mostly all boring mechanical edits. that's the usual problem with using Makefiles directly rather than an actual build generator.

ScottPJones commented:
I like to live on the bleeding edge, so I built out of this branch, did `make clean ; make O=my04`,
but got this near the end:
```
In file included from /j/julia/ui/repl.c:31:
/j/julia/ui/../src/julia.h:1572:10: fatal error: 'julia_version.h' file not found
#include <julia_version.h>
```
`julia_version.h` *is* present in `my04/src`.

vtjnash commented:
> julia_version.h is present in my04/src.

good catch. it should be installed usr/include, but there was no makefile rule for that

tkelman commented:
> it's mostly all boring mechanical edits.

Mostly. Which makes this exceptionally hard to review for the places where significant changes are being made. Please just hold off on this.

vtjnash commented:
> Mostly. Which makes this exceptionally hard to review for the places where significant changes are being made.

you've done a great job already at finding the places that I missed updating a reference. there are also a few build fixes here for a couple of less common build configurations that build upon the initial commit (by my count, at least 8 of these commit are bug fixes, the rest are generally fixes to the initial concept to make the existing out-of-tree build support easier to use)

vtjnash committed:
point ui build at build output directory first, to pickup and build configuration files

ScottPJones commented:
@vtjnash The `=` to `:=` changed seemed completely mechanical, and creates a lot of the "noise", maybe that could go in a separate PR, and that could get merged in without controversy, clearing the air around the meat of your change (which I'd really like to see, if not in `0.4.0`, at the lastest backported to be in `0.4.1`).   I had to chop up my big mega change, but it finally got in.

ScottPJones commented:
@vtjnash Nice, that fixed it!  Now, how well does this play with your codegen changes, and Keno's LLVM SVN fix? :grinning:  (I do want to live on the bleeding edge!)

tkelman commented:
Yeah many of the small bug fix pieces of this would be uncontroversial on their own (though still better as PR's so at least they can get individual review). `:=` still worries me because though it's mechanical and mostly should be equivalent, there's always the possibility of hidden places you haven't found yet (as evidenced by the many rounds of review here, the build system is subtle and has lots of dark corners where things are easily broken) where we were actually relying on delayed expansion. Particularly for distro packagers, who build in unusual configurations and would appreciate some time and warning to test this kind of change as opposed to getting a last-second surprise that they can't package the 0.4.0 release without major patches.

We can think about backporting this after it's stable for a month or two, but how many more people need to say that we shouldn't do this before branching?

ScottPJones commented:
So maybe @vtjnash could split this into a number of PRs of bug-fixes, a PR for `:=` change, and leave this PR with the meat (depending on the `:=` PR for the moment, like I did with my chopping up UTF changes)?  A month or two of stability in `0.5-dev` might work well for getting it backported for a `0.4.1` release? (hoping that `0.4.0` is able to be released in a month or so from now :grinning:)

vtjnash commented:
> @vtjnash Nice, that fixed it! Now, how well does this play with your codegen changes, and Keno's LLVM SVN fix? :grinning: (I do want to live on the bleeding edge!)

should all be independent.

> Yeah many of the small bug fix pieces of this would be uncontroversial on their own (though still better as PR's so at least they can get individual review). := still worries me because though it's mechanical and mostly should be equivalent, there's always the possibility of hidden places you haven't found yet (as evidenced by the many rounds of review here, the build system is subtle and has lots of dark corners where things are easily broken) where we were actually relying on delayed expansion. Particularly for distro packagers, who build in unusual configurations and would appreciate some time and warning to test this kind of change as opposed to getting a last-second surprise that they can't package the 0.4.0 release without major patches.

a lot of the issues you found were generally: (a) places where cygwin had weird compatibility issues (b) i missed a change (c) there was a pre-existing error that was made more obvious by the forced rebuild

> Particularly for distro packagers

seriously, less FUD, please. the essential difference between `=` and `:=` is that `=` effectively creates a function that runs wherever the variable gets used. the documentation doesn't even like them (https://www.gnu.org/software/make/manual/html_node/Flavors.html#Flavors). whereas `:=` creates a variable in the usual sense. there are a couple of places where we use `=` to create functions, but the vast majority of the code is improved by making this change since it's behavior is more predictable in unusual configurations.

we aren't even about to release v0.4-RC1, we are about to enter a period of time to fix the 64 open bugs (https://github.com/JuliaLang/julia/issues?q=is%3Aopen+is%3Aissue+milestone%3A0.4.x)

> a PR for := change,

I had that commit last for a while so you could look at the PR without it, but eventually stopped because the rebase effort to maintain was that split was just wasting time to constantly be removing and re-adding the same fixes during rebasing. git still makes it easy to do so virtually, however. links provided for the curious:
https://github.com/JuliaLang/julia/compare/5505cf8...02be9d9d2a171d7dad6ade6a33901faa89a13d3b
https://github.com/JuliaLang/julia/commit/891b16b8b8e4389655bd6053bba6563adc7d622a
https://github.com/JuliaLang/julia/compare/2ef10f5497d17c1426d29de509f32d3eafeea132...jn/makefile_o

tkelman commented:
The majority of the change here is a feature, not a bug fix, and we are about to declare a feature freeze. I mention distro packagers because so far none of them have reviewed this change or tested that it works in their environment. So far I've been one of the only people doing serious review here. Again, this can wait, I haven't heard a convincing argument yet that this must go into 0.4.

StefanKarpinski commented:
This is not going into 0.4. It could be backported once it proves stable on 0.5 for a while.

vtjnash committed:
disable arpack test entirely since it fails stochastically

ViralBShah commented:
Yes, this is too large a change for 0.4 at this point. Given that 0.5 is expected to have a short release cycle and is not too open ended, perhaps backporting may not also be necessary.

StefanKarpinski commented:
We should backport anything that we can without risking instability or incompatibility on release-0.4, including non-breaking changes to the build system. It's unclear to me if this qualifies but it may.

ViralBShah commented:
It would provide a new API in the build system. Not sure what semver has to say about that.

nalimilan commented:
FWIW, building my RPM package went fine with this branch.

(But I agree this shouldn't be merged before 0.4.)

vtjnash committed:
fix test wildcard rule

vtjnash committed:
cleaup ARCH/MARCH/XC_HOST/BUILD_MACHINE detection logic so it works better & cross-compile works on Darwin

vtjnash committed:
fix llvm-config.exe cross-compile rule detection

vtjnash committed:
add missing updates to := for perf/*/Makefile

vtjnash committed:
add missing win64-int128 patch to llvm3.3 list

vtjnash committed:
fix include location for looking up julia_flisp.boot.inc

vtjnash commented:
bump. this is ready to merge.

tkelman commented:
This should wait for and be rebased on top of #13077. I'd also strongly prefer a less ugly way of dealing with the auto-generated `.jl` files that doesn't mess with `ARGS`.

tkelman commented on doc/Makefile:
```diff
@@ -3,29 +3,35 @@
-JULIA_EXECUTABLE = $(JULIAHOME)/usr/bin/julia
+SPHINXOPTS       :=
+PAPER            :=
+SRCDIR 			 := $(abspath $(dir $(lastword $(MAKEFILE_LIST))))
```
spaces here, not tabs

vtjnash commented:
> This should wait for and be rebased on top of #13077. I'd also strongly prefer a less ugly way of dealing with the auto-generated .jl files that doesn't mess with ARGS.

is that ready to merge soon? otherwise, rebasing a large PR like this on top increases the likelyhood of a mistake, conflict, or overlooked side-effect will introduce an error.

> fatal: attempt to fetch/clone from a shallow repository

this is a an issue with an older git. I can avoid it though if we need to support that.

> edit3: and openlibm has an issue with include paths

can you include a little bit more of the context so I have any idea at all what command line was getting passed to your compiler?

tkelman commented:
> is that ready to merge soon?

It's good to go from me, and needs to be backported unlike this. Could you look at it and merge if you don't see anything wrong with it?

> I can avoid it though if we need to support that.

Don't know which version of git it was, but it was on the buildbots so we do need to support it.

> context

http://buildbot.e.ip.saba.us:8010/builders/package_win6.2-x86/builds/91/steps/make/logs/stdio

vtjnash committed:
also fix contrib/mac/app/Makefile for := change

we have too many Makefiles

vtjnash commented:
thanks. it looks like you added the offending line in https://github.com/JuliaLang/julia/pull/11705/files#diff-3ba529ae517f6b57803af0502f52a40bR792
can you explain the purpose of that change? it probably shouldn't be there according to the design of these Makefiles.

tkelman commented:
IIRC that was an attempt to make building in cygwin work with a non-cygwin toolchain. That managed to fix openlibm, but cmake and libgit2 were less cooperative so that never fully worked. Can probably remove that.

vtjnash commented:
if that was easy, the msys2 project wouldn't need to exist

vtjnash committed:
remove OPENLIBM_HOME override that was breaking the openlibm makefile configuration on cygwin

vtjnash committed:
use spaces not tabs in docs/makefile to lineup :=

tkelman commented:
Yeah, well, it was worth a try. I don't think msys2 really needs to exist as a separate project - if people would be adults about it they'd all just submit patches and there would be an "msys mode" you could start cygwin in.

vtjnash committed:
make the srccache a full git repo

this should make it more practical to develop and hop branches without network access
to cut back down on network traffic, we could consider also switching to NO_GIT tarballs as a default, but that would also prevent incremental recompilation for those dependencies (which could be good or bad depending on your viewpoint), so that change will require more thought

vtjnash commented:
> Don't know which version of git it was, but it was on the buildbots so we do need to support it.

it was added in Git 1.9 (Feb 2014), so sometime older than that apparently. anyways, i've dropped out the need for it in https://github.com/JuliaLang/julia/commit/130083ff0fc6b3290e0bf0868272ccd89df4d725

vtjnash committed:
Merge remote-tracking branch 'origin/master' into jn/makefile_o

vtjnash commented:
> rebased on top of #13077.

git couldn't really handle the fact that that change conflicted with the meaning but not the content of this change, so I had to do a merge commit instead to fix the conflicts. should be good to merge then in the morning, once CI has had a chance to run

tkelman commented:
> less ugly way of dealing with the auto-generated `.jl` files that doesn't mess with `ARGS`.

This is still messy and incomprehensible and I'm not okay with leaving it this way permanently. Are you actually planning on implementing an alternative as discussed in https://github.com/JuliaLang/julia/pull/12463#discussion_r36574931 ?

As it is more people are going to be reading this bootstrapping code and need to debug issues with it than will use out-of-tree builds. Net negative in confusing code vs the build feature addition.

vtjnash commented:
> This is still messy and incomprehensible and I'm not okay with leaving it this way permanently. Are you actually planning on implementing an alternative as discussed in #12463 (comment) ?

yes. but there is absolutely no chance of me making this PR bigger to do it.

tkelman commented:
https://github.com/JuliaLang/julia/pull/12463/files#r39336263

tkelman commented on Makefile:
```diff
@@ -169,40 +198,41 @@ CORE_SRCS := base/boot.jl base/coreimg.jl \
+	@$$(call PRINT_JULIA, cd $$(JULIAHOME)/base && \
 	$$(call spawn,$2) -C $$(JULIA_CPU_TARGET) --output-o $$(call cygpath_w,$$@) $$(JULIA_SYSIMG_BUILD_FLAGS) -f \
-		-J $$(call cygpath_w,$$<) sysimg.jl \
+		-J $$(call cygpath_w,$$<) sysimg.jl $$(RELBUILDROOT) \
```
this probably breaks `build_sysimg.jl`

vtjnash replied:
it shouldn't

tkelman commented:
And merging this should come with a pretty loud announcement to julia-users and/or -dev that everyone who pulls this change will end up rebuilding all deps from scratch, even if they continue building in-tree. Anyone who doesn't want to use out-of-tree builds and plans on doing any work on release-0.4 should probably leave their existing builds in place and start from a separate clone for master once this is merged.

tkelman commented on README.md:
```diff
@@ -77,9 +77,11 @@ When compiled the first time, it will automatically download and build its [exte
 This takes a while, but only has to be done once. If the defaults in the build do not work for you, and you need to set specific make parameters, you can save them in `Make.user`. The build will automatically check for the existence of `Make.user` and use it if it exists.
 Building Julia requires 1.5GiB of disk space and approximately 700MiB of virtual memory.
 
+For release-0.4 and newer builds of Julia, you can create out-of-tree builds of Julia by specifying `make O=<build-directory> configure` on the command line. This will create a directory mirror, with all of the necessary Makefiles to build Julia, in the specified directory. These builds will share the source files in Julia and `deps/srccache`. Each out-of-tree build directory can have its own `Ma ...
```
0.5. If we backport this it won't be for a month or so.

vtjnash committed:
remove changes to PATH variable from WINNT makefile

tkelman doesn't like it

vtjnash committed:
makefile output directory PR did not make it into release-0.4

vtjnash commented:
I'm waiting for CI on master to be green again before merging this. In the meantime, here's my email draft:

vtjnash commented:
I'm waiting for CI on master to be green again before merging this. In the meantime, here's my email draft:
I've recently merged https://github.com/JuliaLang/julia/pull/12463 which enables out-of-tree builds. What this means for you is:

1) deps will rebuild. All source files are now in deps/srccache, all build files are now in deps/build.
2) `make O=output` will start an out-of-tree build in the folder "output"
3) dummy Makefiles are setup in the out-of-tree build folders for easier incremental building from any directory
4) The new makefile target `configure` will create these dummy Makefiles
5) by placing a `Make.user` file at `output/Make.user`, you can override the behavior of the global `Make.user` file for that directory allowing testing for a wider range of systems (windows, 32-bit, llvm-versions, etc), all from the same source tree. Note that the global Make.user file is still included, but is run prior to the Make.user file.
6) unlike most other out-of-tree build systems, this system does not care if you also have an in-tree build

(this is all mentioned in https://github.com/JuliaLang/julia/blob/jn/makefile_o/README.md#source-download-and-compilation also)

tkelman commented:
Cool, there's a new conflict here but +1 from me on merging once that's resolved.

I'd add that if people plan on switching branches to release-0.4, it would be safer to leave existing build trees in place for release-0.4 (to avoid rebuilding) and doing new clones for master. Hopefully the previous issue with switching branches won't recur but it might.

vtjnash commented:
as long as you `rm -rf julia/usr` between switching branches, you should be fine to stay in the same build tree

tkelman commented:
Good call. The issue that I hit earlier with switching between branches with vs without this change looks fixed now.

The buildbots are having some trouble though:

- OSX http://buildbot.e.ip.saba.us:8010/builders/package_osx10.9-x64/builds/644/steps/make%20.app/logs/stdio
- Linux http://buildbot.e.ip.saba.us:8010/builders/package_tarball32/builds/1512/steps/make%20binary-dist/logs/stdio
- Windows http://buildbot.e.ip.saba.us:8010/builders/package_win6.2-x64/builds/122/steps/make%20binary-dist/logs/stdio (there's also an empty ``` `cygpath -w ` ``` in there causing warnings, `RELBUILDROOT` should maybe be `.` instead of empty?)

Keno commented:
Something seems wrong. On a fresh checkout, I get:
```
make: *** No rule to make target 'jltypes.o', needed by '/Users/kfischer/Projects/julia-sa/usr/lib/libjulia.dylib'.  Stop.
```

vtjnash commented:
that means it was missing one of the dependencies, so it couldn't figure out how to build it. i usually see that if libuv.a or llvm-config is missing

Keno commented:
Could we get a better error message for that?

vtjnash commented:
was that the actual issue? i would have to hard-code a test for them, but it is possible.

Keno commented:
It worked fine after `rm -rf`ing `usr`. It doesn't have to be a very specific error message, just telling the user to delete `usr` is enough.

vtjnash commented:
that should have been overkill. `make` at the toplevel should have invoked `make -C deps`, which should have targets to ensure all of the dependencies are installed before it invoked `make -C src`

vtjnash commented:
@tkelman: this seems suspicious to me:
```
make[3]: '/home/Administrator/buildbot/slave/package_win6_2-x64/build/usr/lib/julia/inference.ji' is up to date.
```
why would that file be up-to-date without having built it? is it possible there is old build state stuck around in the form of a file called `usr/.examples` that is inhibiting updating of `usr/share/docs/julia`?

vtjnash commented:
> (there's also an empty `cygpath -w ` in there causing warnings, RELBUILDROOT should maybe be . instead of empty?)

the warning only affects cygwin, and is non-fatal, so i don't care

vtjnash commented:
> Good call. The issue that I hit earlier with switching between branches with vs without this change looks fixed now.

you definitely would still need to `rm -rf usr` to switch very reliably, but the whole point of moving the build to `deps/build` was that it makes that transition cleaner and simpler

tkelman commented:
> why would that file be up-to-date without having built it? is it possible there is old build state stuck around in the form of a file called `usr/.examples` that is inhibiting updating of `usr/share/docs/julia`?

Potentially. The buildbots generally do incremental builds unless we manually clear them out. I just did so, but it'll be a little more time-consuming if every time we go back and forth between building release-0.4 and master we need to completely clean them out. Can probably switch them to doing `rm -rf usr` instead of the `make cleanall` that they're doing now.

vtjnash commented:
maybe not a bad idea to do `rm -rf usr` anyways, but i see the make clean rule that got lost in the rebase attempts.

vtjnash commented:
fixed in 5b40f2bd35d87b267dc1ff1cfc6fd7e07670e89c

mbauman commented:
Did this change the location of JULIA_HOME for local (non-installed) source builds?  I keep editing files in ./usr/share/julia/base instead of ./base.  It's not a big deal, but surely others have run into this, too.  Any workarounds?

vtjnash commented:
usr/share/julia/base is a symlink to base, so the exact path shouldn't matter. the former is a more stable location, however, for working with out-of-tree builds.

mbauman commented:
Should have looked a little closer.  That's great.  Thanks!
